packages/dds/tree/src/feature-libraries/modular-schema/comparison.ts

```diff
@@ -203,6 +213,192 @@ export function allowsRepoSuperset(
 	return true;
 }
 
+/**
+ * @returns the discrepancies between two TreeStoredSchema objects
+ */
+export function allowsDocumentConcepts(
+	original: TreeStoredSchema,
+	updated: TreeStoredSchema,
+): Incompatibility[] {
+	const incompatibilities: Incompatibility[] = [];
+	const areSetsEqual = (a: TreeTypeSet, b: TreeTypeSet): boolean => {
+		if (a === undefined && b === undefined) {
+			return true;
+		}
+		if (a !== undefined && b !== undefined) {
+			return a.size === b.size && [...a].every((value) => b.has(value));
+		}
+		return false;
+	};
+	const getFieldKind = (x: unknown): SchemaFactoryFieldKind | undefined => {
+		switch (x) {
+			case "Value": {
+				return "required";
+			}
+			case "Optional": {
+				return "optional";
+			}
+			case "Sequence": {
+				return "array";
+			}
+			default: {
+				return undefined;
+			}
+		}
+	};
+
+	/**
+	 * The helper function tracks discrepancies between two TreeFieldStoredSchema objects.
+	 */
+	const trackFieldDiscrepancies = (
+		originalField: TreeFieldStoredSchema,
+		updatedField: TreeFieldStoredSchema,
+		key: string,
+	): FieldIncompatibility[] => {
+		const differences: FieldIncompatibility[] = [];
+		if (!areSetsEqual(originalField.types, updatedField.types)) {
+			differences.push({
+				identifier: key,
+				mismatch: "allowedTypes",
+				view: updatedField.types ? [...updatedField.types.values()] : [],
```

Comment:
In practice, this return value will likely not be what code cares about. I think we should make `view` and `stored` contain `view - (view intersect stored)` and `stored - (view intersect stored)`, so that in aggregate they contain the symmetric difference.

This should allow a simpler check for compatibility (since what we really care about is subset relations here): for a given compatibility direction, check that the correct array has size 0.